pkg/controllers/gateway/gateway_controller.go

```diff
@@ -70,88 +88,237 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
-			LastTransitionTime: metav1.Now(),
-			Message:            fmt.Sprintf("Gateways configured in data plane clusters - [%v]", strings.Join(clusters, ",")),
-			Reason:             string(gatewayv1beta1.GatewayConditionProgrammed),
-			Status:             metav1.ConditionTrue,
-			Type:               string(gatewayv1beta1.GatewayConditionProgrammed),
-			ObservedGeneration: previous.Generation,
-		})
-	}
-
-	if !reflect.DeepEqual(gateway.Annotations, previous.Annotations) {
-		log.Info("Updating Gateway annotations", "gateway", gateway)
+	// Update gateway spec/metadata
+	if !reflect.DeepEqual(gateway, previous) {
+		log.Info("Updating Gateway Spec", "gatewaySpec", gateway.Spec, "previousSpec", previous.Spec)
+		log.Info("Updating Gateway ObjectMeta", "gatewayObjectMeta", gateway.ObjectMeta, "previousObjectMeta", previous.ObjectMeta)
 		err = r.Update(ctx, gateway)
 		if err != nil {
-			return ctrl.Result{}, err
+			log.Error(err, "Error updating Gateway")
 		}
 	}
 
-	gateway.Status.Conditions = statusConditions
+	// Update status
+	gateway.Status.Conditions = buildStatusConditions(gateway.Status, previous.Generation, clusters, acceptedStatus, programmedStatus)
 	if !reflect.DeepEqual(gateway.Status, previous.Status) {
-		log.Info("Updating Gateway status", "gatewayStatus", gateway.Status)
+		log.Info("Updating Gateway status", "gatewaystatus", gateway.Status, "previousstatus", previous.Status)
 		err = r.Status().Update(ctx, gateway)
 		if err != nil {
-			return ctrl.Result{}, err
+			log.Error(err, "Error updating Gateway status")
 		}
 	}
 
-	return ctrl.Result{}, nil
+	if requeue {
+		return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, reconcileErr
+	}
+	return ctrl.Result{}, reconcileErr
 }
 
-func applyClusterSyncerAnnotations(gateway *gatewayv1beta1.Gateway, clusters []string) {
-	for _, cluster := range clusters {
-		gateway.Annotations[fmt.Sprintf("%s/%s", ClusterSyncerAnnotation, cluster)] = "True"
+// Configures Gateway tls & dns for each cluster it targets.
+// Returns the programmed status, a list of clusters that were programmed, if the gateway should be requeued, and any error
+func (r *GatewayReconciler) reconcileGateway(ctx context.Context, previous gatewayv1beta1.Gateway, gateway *gatewayv1beta1.Gateway) (metav1.ConditionStatus, []string, bool, error) {
+	log := log.FromContext(ctx)
+
+	clusters := selectClusters(*gateway)
+	// Don't do anything else until at least 1 cluster matches.
+	if len(clusters) == 0 {
+		// TODO: Handle any cleanup if there were previously clusters
+		return metav1.ConditionFalse, clusters, false, nil
 	}
+	trafficAccessor := traffic.NewGateway(gateway)
+	gatewayHelper := trafficAccessor.(GatewayHelper)
+	allHosts := trafficAccessor.GetHosts()
+	hostsWithMZ := []string{}
+
+	// First, filter out hosts that are *not* in a Managed Zone
+	// Those can be left alone
+	for _, host := range allHosts {
+		hostParts := strings.SplitN(host, ".", 2)
```

Comment:
We  might want to use a trie tree structure, example https://github.com/detailyang/domaintree-go. Also I think it might be worth encapsulating in the `GetManagedZoneForDomain` method

Reply:
Good point. Probably worth moving that logic into the method for ease of testing it. Maybe rename it `GetManagedZoneForHost` and just pass the host straight it.

Reply:
The logic btw will work for a co.uk domain, or any host with multiple subdomains, but it is simplistic in that it will only knock off the first label and look for an exactly matching MZ for the rest e.g. foo.bar.baz.cbrooks.co.uk, would look for an MZ for `bar.baz.cbrooks.co.uk` and create a record for `foo` in it.

If we want to be able to work down the tree and for example find an MZ for `cbrooks.co.uk` and create a record for `foo.bar.baz` then we would need to update the logic in some way to allow this.

Reply:
This can be added as a follow up

Reply:
Created a follow up https://github.com/Kuadrant/multi-cluster-traffic-controller/issues/112

Reply:
https://github.com/Kuadrant/multi-cluster-traffic-controller/issues/112